diff -r a0f47af9a7d7 -r 4b63b7e15b58 tools/firmware/hvmloader/acpi/dsdt.asl
--- a/tools/firmware/hvmloader/acpi/dsdt.asl	Wed Jan 26 12:41:39 2011 +0000
+++ b/tools/firmware/hvmloader/acpi/dsdt.asl	Mon Jan 31 14:56:50 2011 +0000
@@ -33,22 +33,22 @@ DefinitionBlock ("DSDT.aml", "DSDT", 2, 
      */
     Name (\_S3, Package (0x04)
     {
-        0x01,  /* PM1a_CNT.SLP_TYP */
-        0x01,  /* PM1b_CNT.SLP_TYP */
+        0x05,  /* PM1a_CNT.SLP_TYP */
+        0x05,  /* PM1b_CNT.SLP_TYP */
         0x0,   /* reserved */
         0x0    /* reserved */
     })
     Name (\_S4, Package (0x04)
     {
-        0x00,  /* PM1a_CNT.SLP_TYP */
-        0x00,  /* PM1b_CNT.SLP_TYP */
+        0x06,  /* PM1a_CNT.SLP_TYP */
+        0x06,  /* PM1b_CNT.SLP_TYP */
         0x00,  /* reserved */
         0x00   /* reserved */
     })
     Name (\_S5, Package (0x04)
     {
-        0x00,  /* PM1a_CNT.SLP_TYP */
-        0x00,  /* PM1b_CNT.SLP_TYP */
+        0x07,  /* PM1a_CNT.SLP_TYP */
+        0x07,  /* PM1b_CNT.SLP_TYP */
         0x00,  /* reserved */
         0x00   /* reserved */
     })
diff -r a0f47af9a7d7 -r 4b63b7e15b58 tools/firmware/hvmloader/acpi/static_tables.c
--- a/tools/firmware/hvmloader/acpi/static_tables.c	Wed Jan 26 12:41:39 2011 +0000
+++ b/tools/firmware/hvmloader/acpi/static_tables.c	Mon Jan 31 14:56:50 2011 +0000
@@ -56,14 +56,14 @@ struct acpi_20_fadt Fadt = {
 
     .sci_int = 9,
 
-    .pm1a_evt_blk = ACPI_PM1A_EVT_BLK_ADDRESS_V1,
-    .pm1a_cnt_blk = ACPI_PM1A_CNT_BLK_ADDRESS_V1,
-    .pm_tmr_blk = ACPI_PM_TMR_BLK_ADDRESS_V1,
-    .gpe0_blk = ACPI_GPE0_BLK_ADDRESS_V1,
+    .pm1a_evt_blk = ACPI_PM1A_EVT_BLK_ADDRESS,
+    .pm1a_cnt_blk = ACPI_PM1A_CNT_BLK_ADDRESS,
+    .pm_tmr_blk = ACPI_PM_TMR_BLK_ADDRESS,
+    .gpe0_blk = ACPI_GPE0_BLK_ADDRESS,
     .pm1_evt_len = ACPI_PM1A_EVT_BLK_BIT_WIDTH / 8,
     .pm1_cnt_len = ACPI_PM1A_CNT_BLK_BIT_WIDTH / 8,
     .pm_tmr_len = ACPI_PM_TMR_BLK_BIT_WIDTH / 8,
-    .gpe0_blk_len = ACPI_GPE0_BLK_LEN_V1,
+    .gpe0_blk_len = ACPI_GPE0_BLK_LEN,
 
     .p_lvl2_lat = 0x0fff, /* >100,  means we do not support C2 state */
     .p_lvl3_lat = 0x0fff, /* >1000, means we do not support C3 state */
@@ -85,21 +85,21 @@ struct acpi_20_fadt Fadt = {
         .address_space_id    = ACPI_SYSTEM_IO,
         .register_bit_width  = ACPI_PM1A_EVT_BLK_BIT_WIDTH,
         .register_bit_offset = ACPI_PM1A_EVT_BLK_BIT_OFFSET,
-        .address             = ACPI_PM1A_EVT_BLK_ADDRESS_V1,
+        .address             = ACPI_PM1A_EVT_BLK_ADDRESS,
     },
 
     .x_pm1a_cnt_blk = {
         .address_space_id    = ACPI_SYSTEM_IO,
         .register_bit_width  = ACPI_PM1A_CNT_BLK_BIT_WIDTH,
         .register_bit_offset = ACPI_PM1A_CNT_BLK_BIT_OFFSET,
-        .address             = ACPI_PM1A_CNT_BLK_ADDRESS_V1,
+        .address             = ACPI_PM1A_CNT_BLK_ADDRESS,
     },
 
     .x_pm_tmr_blk = {
         .address_space_id    = ACPI_SYSTEM_IO,
         .register_bit_width  = ACPI_PM_TMR_BLK_BIT_WIDTH,
         .register_bit_offset = ACPI_PM_TMR_BLK_BIT_OFFSET,
-        .address             = ACPI_PM_TMR_BLK_ADDRESS_V1,
+        .address             = ACPI_PM_TMR_BLK_ADDRESS,
     }
 };
 
diff -r a0f47af9a7d7 -r 4b63b7e15b58 tools/firmware/hvmloader/hvmloader.c
--- a/tools/firmware/hvmloader/hvmloader.c	Wed Jan 26 12:41:39 2011 +0000
+++ b/tools/firmware/hvmloader/hvmloader.c	Mon Jan 31 14:56:50 2011 +0000
@@ -31,7 +31,6 @@
 #include "option_rom.h"
 #include <xen/version.h>
 #include <xen/hvm/params.h>
-#include <xen/hvm/ioreq.h>
 #include <xen/memory.h>
 
 asm (
@@ -223,12 +222,9 @@ static void pci_setup(void)
             /* PIIX4 ACPI PM. Special device with special PCI config space. */
             ASSERT((vendor_id == 0x8086) && (device_id == 0x7113));
             pci_writew(devfn, 0x20, 0x0000); /* No smb bus IO enable */
-            pci_writew(devfn, 0xd2, 0x0000); /* No smb bus IO enable */
             pci_writew(devfn, 0x22, 0x0000);
             pci_writew(devfn, 0x3c, 0x0009); /* Hardcoded IRQ9 */
             pci_writew(devfn, 0x3d, 0x0001);
-            pci_writel(devfn, 0x40, ACPI_PM1A_EVT_BLK_ADDRESS_V1 | 1);
-            pci_writeb(devfn, 0x80, 0x01); /* enable PM io space */
             break;
         case 0x0101:
             if ( vendor_id == 0x8086 )
@@ -781,12 +777,12 @@ int main(void)
         struct xen_hvm_param p = {
             .domid = DOMID_SELF,
             .index = HVM_PARAM_ACPI_IOPORTS_LOCATION,
-            .value = 1,
+            .value = 0,
         };
 
+        hypercall_hvm_op(HVMOP_set_param, &p);
         printf("Loading ACPI ...\n");
         acpi_build_tables();
-        hypercall_hvm_op(HVMOP_set_param, &p);
     }
 
     init_vm86_tss();
