# HG changeset patch
# Parent d5764e24f647f39846a54861b1a7b9b8246afb65

diff -r d5764e24f647 xen/arch/x86/cpu/intel.c
--- a/xen/arch/x86/cpu/intel.c
+++ b/xen/arch/x86/cpu/intel.c
@@ -107,7 +107,7 @@ static void __devinit set_cpuidmask(cons
 			return;
 		extra = "xsave ";
 		break;
-	case 0x2a:
+	case 0x2a: case 0x2d:
 		wrmsr(MSR_INTEL_CPUID1_FEATURE_MASK_V2,
 		      opt_cpuid_mask_ecx,
 		      opt_cpuid_mask_edx);
